### ISSUE TITLE
1st Esc in completions file finder exits vi-mode's insert, 2nd exits completions

### ISSUE THREAD

**Issue by mqudsi** [fish-shell/fish-shell#2871]
Running off of master, I believe there is a bug in the behavior of the Esc key when browsing file completions with vi-mode in insert.
### Reproduction Steps:
1. Enter vi-mode
2. Enter insert mode
3. Trigger the file finder feature and tab through the results, stopping midway
4. Press Esc
5. Press Esc
### Expected behavior:

The first Esc should exit the file finder, not vi-mode insert. The second Esc should exit vi-mode insert.
### Observed behavior:

The first Esc exits vi-mode while keeping you in the file browser.
### Additional information:

---

Fish version: fish, version 2.2.0-708-gd30f8ff

Operating system: OS X, installed from git/master

Terminal or terminal emulator: OS X terminal

**Comment by faho** [comment]
This isn't actually a regression, being present in 2.2.0 - the first release to feature vi-mode.

What it is is more complicated to fix than I initially assumed. The thing is that escape in insert mode never executes `commandline -f cancel`. This could be fixed by adding it, either conditionally if `commandline --paging-mode` is true or unconditionally - I'm not sure what `cancel` actually cancels except for paging, so I'm not sure if there's any harm in that.

BUT: What I'd imagine this should do is to exit the pager if there is, or switch to default mode otherwise. And that doesn't seem possible currently.

As always with vi-mode, I'd appreciate feedback from those who use it.

**Comment by christhekeele** [comment]
To weigh in, this definitely is a little irritating. The experience in vi mode is:

```diff
[type text]    -> start command
[tab]          -> look through autocomplete
[type text]    -> search autocomplete
[esc]          -> decide to refine your command
---
< [esc]        -> because you're still stuck in the pager, but now in normal mode, you esc again
< [can't type] -> to immediately refine your command
---
> [type text]  -> to immediately refine your command
> [esc]        -> at leisure to go to normal mode and manipulate refined command
---
```

**Comment by ridiculousfish** [comment]
PR #3799 is my attempt at fixing this. I appreciate any testing or feedback before merging.

**Comment by ridiculousfish** [comment]
Fixed in 8a09c99f36977ec98ebbc7c1e5255402da5dbaa8
